amd_tressfx_vulkan/src/TressFXSimulationVulkan.cpp

```diff
@@ -0,0 +1,1180 @@
+//
+// Simulate
+//
+// Runs the hair simulation which will animate the hair vertices. This function calls
+// Dispatch to execute compute shaders that implement the simulation.
+//
+////--------------------------------------------------------------------------------------
+VkResult TressFXSimulation::Simulate(VkDevice pvkDevice, VkCommandBuffer commandBuffer,
+                                     float fElapsedTime, float density,
+                                     tressfx_vec3 &windDir, float windMag,
+                                     XMMATRIX *pModelTransformForHead,
+                                     ID3D11UnorderedAccessView *pSkinningTransforms,
+                                     float targetFrameRate, bool singleHeadTransform,
+                                     bool warp, uint32_t uniformBufferIndex)
+{
+    (void)pSkinningTransforms;
+    m_elapsedTimeSinceLastSim += fElapsedTime;
+    bool bFullSimulate = true;
+
+    // Simulation is sensitive to frame rate. So, we set the target frame rate (defaulted
+    // to 1/60)
+    // and if the current elapsed time since the last simulation is too early,
+    // we run the simulation with lower iterations and stiffness values.
+    if ((m_elapsedTimeSinceLastSim < targetFrameRate) && !warp)
+    {
+        bFullSimulate = false;
+    }
+    else
+    {
+        m_elapsedTimeSinceLastSim = 0;
+    }
+
+    VkResult vr = VK_SUCCESS;
+
+    int numOfStrandsPerThreadGroup =
+        THREAD_GROUP_SIZE / m_pTressFXMesh->m_HairAsset.m_NumOfVerticesInStrand;
+
+    ConstBufferCS_Per_Frame *pCSPerFrame;
+    vkMapMemory(pvkDevice, m_pCBCSPerFrameMemory,
+                uniformBufferIndex * sizeof(ConstBufferCS_Per_Frame),
+                sizeof(ConstBufferCS_Per_Frame), 0,
+                reinterpret_cast<void **>(&pCSPerFrame));
+
+    {
+        pCSPerFrame->bWarp = warp;
+
+        if (bFullSimulate)
+        {
+            pCSPerFrame->NumLengthConstraintIterations =
+                m_simParams.numLengthConstraintIterations;
+        }
+        else
+        {
+            pCSPerFrame->NumLengthConstraintIterations = 1;
+        }
+
+        pCSPerFrame->bCollision = (m_simParams.bCollision == true) ? 1 : 0;
+
+        pCSPerFrame->GravityMagnitude = m_simParams.gravityMagnitude;
+
+        pCSPerFrame->timeStep = targetFrameRate;
+
+        pCSPerFrame->NumOfStrandsPerThreadGroup = numOfStrandsPerThreadGroup;
+        pCSPerFrame->NumFollowHairsPerGuideHair =
+            (m_bGuideFollowHairPrev
+                 ? m_pTressFXMesh->m_HairAsset.m_NumFollowHairsPerGuideHair
+                 : 0);
+        pCSPerFrame->TipSeparationFactor =
+            m_pTressFXMesh->m_HairAsset.m_TipSeparationFactor;
+
+        ComputeWindPyramid(pCSPerFrame->Wind, pCSPerFrame->Wind1, pCSPerFrame->Wind2,
+                           pCSPerFrame->Wind3, windMag, windDir);
+
+        int numSections = m_simParams.numHairSections;
+
+        // hair section 0
+        if (numSections > 0)
```

Comment:
Switch with fall-through seems to make this statement simpler. Also, I'd move it to a separate function wholesale.